crates/bevy_ecs/src/observer/builder.rs

```diff
@@ -0,0 +1,233 @@
+use std::any::TypeId;
+
+use super::*;
+
+/// Builder struct for [`Observer`].
+pub struct ObserverBuilder<'w, E = ()> {
+    commands: Commands<'w, 'w>,
+    descriptor: ObserverDescriptor,
+    _marker: PhantomData<E>,
+}
+
+impl<'w, E: 'static> ObserverBuilder<'w, E> {
+    /// Constructs a new [`ObserverBuilder`].
+    pub fn new(commands: Commands<'w, 'w>) -> Self {
+        let mut descriptor = ObserverDescriptor::default();
+        let event = commands
+            .components()
+            .get_id(TypeId::of::<E>())
+            .unwrap_or_else(|| {
+                panic!(
+                    "Cannot observe event before it is registered with init_component: {}",
+                    std::any::type_name::<E>(),
+                )
+            });
+        descriptor.events.push(event);
+
+        Self {
+            commands,
+            descriptor,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Constructs an [`ObserverBuilder`] with a dynamic event id.
+    /// # Safety
+    /// Caller must ensure that the component associated with `id` is accessible as E
+    #[must_use]
+    pub unsafe fn new_with_id(event: ComponentId, commands: Commands<'w, 'w>) -> Self {
+        let mut descriptor = ObserverDescriptor::default();
+        descriptor.events.push(event);
+
+        Self {
+            commands,
+            descriptor,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Adds `NewE` to the list of events listened to by this observer.
+    /// Observers that listen to multiple types of events can no longer access the typed event data.
+    pub fn on_event<NewE: 'static>(&mut self) -> &mut ObserverBuilder<'w, ()> {
+        let event = self
+            .commands
+            .components()
+            .get_id(TypeId::of::<NewE>())
+            .unwrap_or_else(|| {
+                panic!(
+                    "Cannot observe event before it is registered with init_component: {}",
+                    std::any::type_name::<NewE>(),
+                )
+            });
+        self.descriptor.events.push(event);
+        // SAFETY: () will not allow bad memory access as it has no size
+        unsafe { std::mem::transmute(self) }
+    }
+
+    /// Add `events` to the list of events listened to by this observer.
+    /// Observers that listen to multiple types of events can no longer access the typed event data.
+    pub fn on_event_ids(
+        &mut self,
+        events: impl IntoIterator<Item = ComponentId>,
+    ) -> &mut ObserverBuilder<'w, ()> {
+        self.descriptor.events.extend(events);
+        // SAFETY: () type will not allow bad memory access as it has no size
+        unsafe { std::mem::transmute(self) }
+    }
+
+    /// Add [`ComponentId`] in `T` to the list of components listened to by this observer.
+    pub fn components<B: Bundle>(&mut self) -> &mut Self {
```

Comment:
It's unclear from this documentation whether components listened to by the observer behave in a similar way to events where you lose access to the data if you listen to multiple. Especially because events are components.

Reply:
You don't get access to any data about the component other than it's ID so the component set is mostly internal. If you want to actually access the component you need to do it via a world reference or a query (providing a pointer gets gnarly for aliasing), but you're right this should be more clear, I think there are many places where the docs need fleshing out. There's also currently no way to get all the component ids if you have an observer that listens for `OnAdd` on multiple components and then all of them are added simultaneously (after the first match for an event observers aren't fired again).

The whole events being components thing is a bit unfortunate as it's confusing to communicate when the "component" isn't actually ever attached to any entity, but building out an entirely new API for events when both are just "type registered to the ECS" seems overkill. We could hide this better like we do for resources, it just bloats out the PR more.

Reply:
It's not fully clear to me how `components` work, I haven't ran into any direct explanation.

If I understand correctly, `components` allows us to target `OnAdd`, `OnRemove` events on specific component-sets, but how do they work with custom events?

Reply:
When `OnAdd` is triggered it is triggered for a specific component on a specific entity i.e. entity X has component A added. So when creating an event you need some way to choose which components you are targeting (if any), that's what `.components` does.

A given custom event doesn't need to be targeted at components, you can just target entities and any observers that are listening for that event will trigger, unless they are only listening for events that target a specific component.

This is a deviation from the flecs API in that flecs always requires you to target a component and all observers have to be listening to events targeting a specific set of components, we could do that as well in order to enforce the consistency of always having component targets, but on the other hand many users won't care about the component they are targeting and just want to send an event at an entity.